File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    int deliver_allowance = 0; // 남은 배달 가능량
    int pickup_allowance = 0;  // 남은 수거 가능량

    // 가장 먼 집부터 거꾸로 확인
    for (int i = n - 1; i >= 0; i--) {
        int cnt = 0; // 해당 집을 방문해야 하는 횟수

        // 현재 집에 처리할 물량이 남아있는지 확인 (이전 방문의 여유분으로 처리 가능한지)
        deliver_allowance -= deliveries[i];
        pickup_allowance -= pickups[i];

        // 여유분이 음수가 되면, 그만큼 더 방문해서 처리해야 함
        while (deliver_allowance < 0 || pickup_allowance < 0) {
            deliver_allowance += cap;
            pickup_allowance += cap;
            cnt++;
        }

        // 방문 횟수 * 거리 * 2(왕복)를 더함
        if (cnt > 0) {
            answer += (long long)(i + 1) * cnt * 2;
        }
    }

    return answer;
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
            
        int del = n-1; // 배달이 남은 가장 먼 집
        int pic = n-1; // 픽업이 남은 가장 먼집
        
        // 가장 먼 집부터 배달, 픽업
        while(true){ // 10만 * 100 = 1억
            // 배달 첫 집
            while(del >= 0 && deliveries[del] == 0) del--;
            // 픽업 첫 집
            while(pic >= 0 && pickups[pic] == 0) pic--;
            
            // 둘다 -1이 되어 끝났다면 끝내기
            if(del < 0 && pic < 0) break;
                        
            // 이동거리 갱신
            answer += (Math.max((del < 0) ? -1 : del, (pic < 0) ? -1 : pic) + 1) * 2;    
            
            // 배달, 수거 가능 크기 확보
            int dc = cap; int pc = cap; 
            // 배달
            while(dc > 0 && del >= 0){ // 50 - 공간도 있고 할 것도 남아있어
                if(deliveries[del] > 0){
                    dc--; deliveries[del]--; 
                }
                if(deliveries[del] <= 0){
                    del--; 
                }
            }
            // 수거
            while(pc > 0 && pic >= 0){ // 50 - 공간도 있고 할 것도 남아있어
                if(pickups[pic] > 0){ 
                    pc--; pickups[pic]--; 
                }
                if(pickups[pic] <= 0) {
                    pic--; // 여기서 pic가 음수가 될 수 있음
                }
            }
            // 물류창고로 가기
        }
        
        return answer;
    }
}